### ISSUE TITLE
Kubernetes storage fails to read namespace from jwt and crashloops pod

### ISSUE THREAD

**Issue by betermieux** [dexidp/dex#2082]
#### Expected behavior (what you expected to happen)
Dex should not Crashloop

#### Actual behavior (what actually happened)
Dex Crashloops
Logs indicate a problem with the namespace detection from service account token
```
failed to initialize storage: failed to inspect service account token: jwt claim "kubernetes.io/serviceaccount/namespace" not found
```
#### Environment

- Dex version: 2.28.0
- Storage type: kubernetes
- Installation type: Helm chart 0.0.7

#### Anything else we should know?

Problem appeared after I upgraded kubernetes to v1.21.0

I have checked the jwt from the service account secret:
```
kubectl -n auth get secret dex-token-r8bgw -o jsonpath='{.data.token}' | base64 --decode | jwt decode -
...
{
  "iss": "kubernetes/serviceaccount",
  "kubernetes.io/serviceaccount/namespace": "auth",
  "kubernetes.io/serviceaccount/secret.name": "dex-token-r8bgw",
  "kubernetes.io/serviceaccount/service-account.name": "dex",
  "sub": "system:serviceaccount:auth:dex"
}
...

```

This seems to be OK, since dex looks for the claim "kubernetes.io/serviceaccount/namespace".
But when I check the token from inside the container, I get strange results
```
cat /run/secrets/kubernetes.io/serviceaccount/token | jwt decode -
{
...
  "iss": "https://kubernetes.default.svc.cluster.local",
  "kubernetes.io": {
    "namespace": "auth",
    "pod": {
      "name": "dex-6f9bc8769b-dmcvg"
    },
    "serviceaccount": {
      "name": "dex"
    }
  },
  "sub": "system:serviceaccount:auth:dex"
}
...
```
This seems to be roughly the same data, but the claims are named differently. Any ideas why the token has changed? And while I am asking: Why doesn't dex use the provided /run/secrets/kubernetes.io/serviceaccount/namespace file? This seems to be less error prone then parsing the jwt token.

**Comment by nabokihms** [comment]
Hello! Thanks for reporting. I don't think anyone has tested Dex with Kubernetes v1.21 yet. We definitely need to work on our Kubernetes Dex testing environment to make testing new Kubernetes releases easier.

About the problem:
Could you try to use downward API as a workaround? I assume that adding the following section to the pod's env will solve the problem.
 ```yaml
- name: KUBERNETES_POD_NAMESPACE
   valueFrom:
    fieldRef:
      fieldPath: metadata.namespace
```

**Comment by nabokihms** [comment]
Oh, it happened because `BoundServiceAccountTokenVolume` in Kubernetes v1.21 is enabled by default.
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.21.md#csi-service-account-token

> The BoundServiceAccountTokenVolume feature has been promoted to beta, and enabled by default.
This changes the tokens provided to containers at /var/run/secrets/kubernetes.io/serviceaccount/token to be time-limited, auto-refreshed, and invalidated when the containing pod is deleted.

**Comment by betermieux** [comment]
Hi! 
Thanks for the tip, after adding the environment variable to the pod Dex works like expected on kubernetes v1.21.
The underlying bug still needs to be fixed, consider using /run/secrets/kubernetes.io/serviceaccount/namespace directly without parsing jwt.
And you are right regarding `BoundServiceAccountTokenVolume`, I have read the same section a few hours ago, but dismissed it because it was mentioning CSI. After reading it again, the jwt claims from the pod make perfect sense.

**Comment by betermieux** [comment]
Link to kubernetes docs regarding `/run/secrets/kubernetes.io/serviceaccount/namespace`:
https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/#accessing-the-api-from-a-pod
The namespace file seems to available since a long time, but I couldn't find an actual kubernetes release which introduced it.

**Comment by nabokihms** [comment]
I opened a PR with the fix for namespace discovery, but this is only half of the problem. We also need to provide a reloading mechanism for tokens, the same as client-go does.

**Comment by 15077693d** [comment]
> ```yaml
> KUBERNETES_POD_NAMESPACE
> ```

Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? Thank you. If yes which part does it belong to ? etc spec, status.

**Comment by 15077693d** [comment]
> Hello! Thanks for reporting. I don't think anyone has tested Dex with Kubernetes v1.21 yet. We definitely need to work on our Kubernetes Dex testing environment to make testing new Kubernetes releases easier.
> 
> About the problem:
> Could you try to use downward API as a workaround? I assume that adding the following section to the pod's env will solve the problem.
> 
> ```yaml
> - name: KUBERNETES_POD_NAMESPACE
>   valueFrom:
>    fieldRef:
>      fieldPath: metadata.namespace
> ```
Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? Thank you. If yes which part does it belong to ? etc spec, status.

**Comment by 15077693d** [comment]
> Hi!
> Thanks for the tip, after adding the environment variable to the pod Dex works like expected on kubernetes v1.21.
> The underlying bug still needs to be fixed, consider using /run/secrets/kubernetes.io/serviceaccount/namespace directly without parsing jwt.
> And you are right regarding `BoundServiceAccountTokenVolume`, I have read the same section a few hours ago, but dismissed it because it was mentioning CSI. After reading it again, the jwt claims from the pod make perfect sense.

Should I set BoundServiceAccountTokenVolume to false to solve the problem? If yes, How to set it? Thank you.

**Comment by betermieux** [comment]
> Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? If yes which part does it belong to ? etc spec, status. Thank you.

it should be put in the dex container environment of the pod:

```
spec:
  containers:
    - env:
      - name: KUBERNETES_POD_NAMESPACE
        …
```

If you are using helm you can set it in `values.yaml` at `env:`
You don't need to disable BoundServiceAccountTokenVolume

**Comment by 15077693d** [comment]
> > Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? If yes which part does it belong to ? etc spec, status. Thank you.
> 
> it should be put in the dex container environment of the pod:
> 
> ```
> spec:
>   containers:
>     - env:
>       - name: KUBERNETES_POD_NAMESPACE
>         …
> ```
> 
> If you are using helm you can set it in `values.yaml` at `env:`
> You don't need to disable BoundServiceAccountTokenVolume
Can i edit it on the kubernetes dashboard? I didn't have helm.

**Comment by 15077693d** [comment]
> > Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? If yes which part does it belong to ? etc spec, status. Thank you.
> 
> it should be put in the dex container environment of the pod:
> 
> ```
> spec:
>   containers:
>     - env:
>       - name: KUBERNETES_POD_NAMESPACE
>         …
> ```
> 
> If you are using helm you can set it in `values.yaml` at `env:`
> You don't need to disable BoundServiceAccountTokenVolume

Could i edit it on the kubernetes dashboard? I didn't have helm.

**Comment by betermieux** [comment]
Sure, open the deployment an edit/add
```
spec:
  template:
    spec:
      containers:
        - name:dex
          env:
          - name: KUBERNETES_POD_NAMESPACE
            value: your-namespace-name
```

**Comment by 15077693d** [comment]
> > Hi nabokihms, Where should i put this section? It is on the auth pod yaml file? If yes which part does it belong to ? etc spec, status. Thank you.
> 
> it should be put in the dex container environment of the pod:
> 
> ```
> spec:
>   containers:
>     - env:
>       - name: KUBERNETES_POD_NAMESPACE
>         …
> ```
> 
> If you are using helm you can set it in `values.yaml` at `env:`
> You don't need to disable BoundServiceAccountTokenVolume

Thank you so much!!!!!!!!!!!!!!!! I solved it!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by mat-ej** [comment]
> spec:
>   template:
>     spec:
>       containers:
>         - name:dex
>           env:
>           - name: KUBERNETES_POD_NAMESPACE
>             value: your-namespace-name

I've tried this and it did not work.

**Comment by betermieux** [comment]
@mat-ej: this is just a snippet of the deployment YAML. It should work, but you have to leave the other lines untouched. And mind the indentation!

**Comment by nabokihms** [comment]
Fixed in v2.30.0

**Comment by themowski** [comment]
Our team encountered this while deploying Kubeflow v1.3.1 via manifests in a Kubernetes v1.21.9 cluster. The solution in our case was to instead install Kubeflow 1.4.x via manifests, which includes the patch https://github.com/kubeflow/manifests/pull/1883 mentioned above.
